troubleshoot-lock-conflicts.md

```diff
@@ -0,0 +1,233 @@
+---
+title: Troubleshoot Lock Conflicts
+summary: Learn to analyze and resolve lock conflicts in TiDB.
+---
+
+# Troubleshoot Lock Conflicts
+ 
+TiDB supports complete distributed transactions. Starting from v3.0, TiDB provides optimistic transaction mode and pessimistic transaction mode. This document introduces how to troubleshoot and resolve lock conflicts in TiDB.
+
+## Optimistic transaction mode
+
+Transactions in TiDB use two-phase commit (2PC) that includes the Prewrite phase and the Commit phase. The procedure is as follows:
+
+![two-phase commit in the optimistic transaction mode](/media/troubleshooting-lock-pic-01.png)
+ 
+For details of Percolator and TiDB’s algorithm of the transactions, see [Google's Percolator](https://ai.google/research/pubs/pub36726).
+
+### Prewrite phase (optimistic)
+ 
+In the Prewrite phase, TiDB adds a primary lock and a secondary lock to target keys. If there are lots of requests for adding locks from the clients, TiDB prints an error such as write conflict or `keyislocked` to the log and reports it to the client. Specifically, the following errors related to locks might occur in the Prewrite phase.
+ 
+#### Read-write conflict (optimistic)
+ 
+As the TiDB server receives a read request from a client, it gets a globally unique and increasing timestamp at the physical time as the start_ts of the current transaction. The transaction needs to read the latest data before start_ts, that is, the target key of the latest commit_ts that is smaller than start_ts. When the transaction finds that the target key is locked by another transaction, and it cannot know which phase the other transaction is in, a read-write conflict happens. The diagram is as follows:
+
+![read-write conflict](/media/troubleshooting-lock-pic-04.png)
+
+Txn0 completes the Prewrite phase and enters the Commit phase. At this time, Txn1 requests to read the same target key. Txn1 needs to read the target key of the latest commit_ts that is smaller than its start_ts. Because Txn1’s start_ts is smaller than Txn0's lock_ts, Txn1 must wait for the target key's lock to be cleaned, but it hasn’t been done. As a result, Txn1 cannot confirm whether Txn0 has been committed or not. Thus, a read-write conflict between Txn1 and Txn0 happens.
```

Comment:
```suggestion
Txn0 completes the Prewrite phase and enters the Commit phase. At this time, Txn1 requests to read the same target key. Txn1 needs to read the target key of the latest commit_ts that is smaller than its start_ts. Because Txn1’s start_ts is smaller than Txn0's lock_ts, Txn1 must wait for the target key's lock to be cleared, but it hasn’t been done. As a result, Txn1 cannot confirm whether Txn0 has been committed or not. Thus, a read-write conflict between Txn1 and Txn0 happens.
```